src/client/index.ts

```diff
@@ -1,12 +1,73 @@
+ *
+ * Notifications will be displayed as toasts as well as being listed and organized in a Notification Center. The
+ * Notification Center can be accessed by clicking on the system tray notification icon.
+ *
+ * ## Getting Started
+ *
+ * See the [Desktop Services documentation](https://developers.openfin.co/docs/desktop-services) for details on
+ * including a desktop service within your application. Once configured via your application manifest, the API
+ * documented here will function as expected.
+ *
+ * ### Basic example:
+ * ```ts
+ * // Non-Interactive "fire and forget" notification
+ * // Clicking the notification will pass an `notification-activated` event back to the application.
+ * // There will still be a `notification-closed` event when the service closes the notification (this occurs in all cases).
+ * await create({
+ *     title: 'Build Complete',
+ *     body: 'Job "develop#123" finished with state "SUCCESS"',
+ *     category: 'Build Statuses'
+ * });
+ * ```
+ *
+ * @module Notifications
  */
 
-import {tryServiceDispatch, eventEmitter} from './connection';
-import {APITopic} from './internal';
+/**
+ * Need a comment block here so that the comment block above is interpreted as a file comment, and not a comment on the
+ * import below.
+ *
+ * @hidden
+ */
+import {ActionDeclaration, NotificationActionResult} from './actions';
+import {tryServiceDispatch, eventEmitter, getEventRouter} from './connection';
+import {ButtonOptions, ControlOptions} from './controls';
+import {APITopic, Events, NotificationInternal} from './internal';
+import {EventRouter, EventTransport} from './EventRouter';
+
+const eventHandler: EventRouter<Events> = getEventRouter();
 
-import {NotificationOptions, Notification, NotificationClickedEvent, NotificationClosedEvent, NotificationButtonClickedEvent} from './index';
-import {NotificationEvent} from './index';
+function parseEventWithNotification<T extends {notification: NotificationInternal}>(event: T): T & {notification: Notification} {
+    const {notification} = event;
+
+    return {
+        ...event,
+        notification: {
+            ...notification,
+            date: new Date(notification.date)
+        }
+    };
+}
+
+eventHandler.registerDeserializer<NotificationCreatedEvent>('notification-created', (event: EventTransport<Events, NotificationCreatedEvent>) => {
+    return parseEventWithNotification(event);
+});
+eventHandler.registerDeserializer<NotificationClosedEvent>('notification-closed', (event: EventTransport<Events, NotificationClosedEvent>) => {
+    return parseEventWithNotification(event);
+});
+eventHandler.registerDeserializer<NotificationActionEvent>('notification-action', (event: EventTransport<Events, NotificationActionEvent>) => {
+    const {controlSource, controlIndex, target, ...rest} = parseEventWithNotification(event);
+
+    if (event.trigger === ActionTrigger.CONTROL && controlSource && controlIndex !== undefined) {
```

Comment:
The type-checks were more to satisfy TypeScript, since the fields are declared as optional.

Could switch to `event.notification[controlSource!][controlIndex!]` instead?

Not really sure about the exception handling here - I'd prefer to rely on a 'design by contract' approach that basically just assumes that can't happen and so not worth adding the bloat to handle it... But I dunno, maybe it could? Down the road?..

Reply:
I prefer `event.notification[controlSource!][controlIndex!]`